// arui-scripts/build/configs/dev-server.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const path_1 = __importDefault(require("path"));
const app_configs_1 = __importDefault(require("./app-configs"));
const apply_overrides_1 = __importDefault(require("./util/apply-overrides"));
const devServerConfig = apply_overrides_1.default('devServer', {
    port: app_configs_1.default.clientServerPort,
    liveReload: false,
    client: {
        overlay: true,
    },
    devMiddleware: {
        publicPath: `/${app_configs_1.default.publicPath}`,
    },
    static: [app_configs_1.default.serverOutputPath],
    proxy: Object.assign(app_configs_1.default.appPackage.proxy || {}, {
        '/**': {
            target: `http://localhost:${app_configs_1.default.serverPort}`,
            bypass: (req) => {
                var _a;
                const assetsRoot = path_1.default.normalize(`/${app_configs_1.default.publicPath}`).replace(/\\/g, '/');
                if ((_a = req === null || req === void 0 ? void 0 : req.url) === null || _a === void 0 ? void 0 : _a.startsWith(assetsRoot)) {
                    return req.url;
                }
                return null;
            }
        }
    })
});
exports.default = devServerConfig;
