// arui-scripts/build/configs/webpack.server.prod.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs_1 = __importDefault(require("fs"));
const path_1 = __importDefault(require("path"));
const webpack_1 = __importDefault(require("webpack"));
const webpack_node_externals_1 = __importDefault(require("webpack-node-externals"));
const fork_ts_checker_webpack_plugin_1 = __importDefault(require("fork-ts-checker-webpack-plugin"));
const tsconfig_paths_webpack_plugin_1 = __importDefault(require("tsconfig-paths-webpack-plugin"));
const getCSSModuleLocalIdent_1 = __importDefault(require("react-dev-utils/getCSSModuleLocalIdent"));
const app_configs_1 = __importDefault(require("./app-configs"));
const babel_server_1 = __importDefault(require("./babel-server"));
const postcss_1 = __importDefault(require("./postcss"));
const apply_overrides_1 = __importDefault(require("./util/apply-overrides"));
const get_entry_1 = __importDefault(require("./util/get-entry"));
const assetsIgnoreBanner = fs_1.default.readFileSync(require.resolve('./util/node-assets-ignore'), 'utf8');
const sourceMapSupportBanner = fs_1.default.readFileSync(require.resolve('./util/install-sourcemap-support'), 'utf8');
// style files regexes
const cssRegex = /\.css$/;
const cssModuleRegex = /\.module\.css$/;
function getSingleEntry(entryPoint) {
    return [...entryPoint];
}
// This is the production configuration.
// It compiles slowly and is focused on producing a fast and minimal bundle.
// The development configuration is different and lives in a separate file.
const config = apply_overrides_1.default(['webpack', 'webpackServer', 'webpackProd', 'webpackServerProd'], {
    mode: 'production',
    // Don't attempt to continue if there are any errors.
    bail: true,
    // We generate sourcemaps in production. This is slow but gives good results.
    devtool: 'source-map',
    target: 'node',
    node: {
        __filename: true,
        __dirname: true
    },
    entry: get_entry_1.default(app_configs_1.default.serverEntry, getSingleEntry),
    context: app_configs_1.default.cwd,
    output: {
        // Add /* filename */ comments to generated require()s in the output.
        pathinfo: true,
        path: app_configs_1.default.serverOutputPath,
        publicPath: app_configs_1.default.publicPath,
        filename: app_configs_1.default.serverOutput,
        chunkFilename: '[name].js',
        // Point sourcemap entries to original disk location (format as URL on Windows)
        devtoolModuleFilenameTemplate: (info) => path_1.default
            .relative(app_configs_1.default.appSrc, info.absoluteResourcePath)
            .replace(/\\/g, '/'),
    },
    externalsPresets: { node: true },
    externals: [webpack_node_externals_1.default({
            allowlist: [
                /^arui-feather/,
                /^arui-ft-private/,
                /^arui-private/,
                /^alfaform-core-ui/,
                /^newclick-components/,
                /^newclick-composite-components/,
                /^#/,
                /^@alfalab\/icons/,
                /^@alfalab\/core-components/,
                /^date-fns/,
            ],
            // we cannot determine node_modules location before arui-scripts installation, so just load
            // dependencies list from package.json
            modulesFromFile: true,
        })],
    resolve: {
        // This allows you to set a fallback for where Webpack should look for modules.
        // We placed these paths second because we want `node_modules` to "win"
        // if there are any conflicts. This matches Node resolution mechanism.
        // https://github.com/facebookincubator/create-react-app/issues/253
        modules: ['node_modules', app_configs_1.default.appNodeModules],
        // These are the reasonable defaults supported by the Node ecosystem.
        // We also include JSX as a common component filename extension to support
        // some tools, although we do not recommend using it, see:
        // https://github.com/facebookincubator/create-react-app/issues/290
        // `web` extension prefixes have been added for better support
        // for React Native Web.
        extensions: ['.web.js', '.mjs', '.js', '.json', '.web.jsx', '.jsx', '.ts', '.tsx'],
        plugins: ([
            (app_configs_1.default.tsconfig && new tsconfig_paths_webpack_plugin_1.default({
                configFile: app_configs_1.default.tsconfig,
                extensions: ['.web.js', '.mjs', '.js', '.json', '.web.jsx', '.jsx', '.ts', '.tsx']
            }))
        ].filter(Boolean)),
    },
    module: {
        // typescript interface will be removed from modules, and we will get an error on correct code
        // see https://github.com/webpack/webpack/issues/7378
        strictExportPresence: !app_configs_1.default.tsconfig,
        rules: [
            {
                oneOf: ([
                    // Process JS with Babel.
                    {
                        test: app_configs_1.default.useTscLoader ? /\.(js|jsx|mjs)$/ : /\.(js|jsx|mjs|ts|tsx)$/,
                        include: app_configs_1.default.appSrc,
                        loader: require.resolve('babel-loader'),
                        options: babel_server_1.default,
                    },
                    (app_configs_1.default.tsconfig && app_configs_1.default.useTscLoader) && {
                        test: /\.tsx?$/,
                        use: [
                            {
                                loader: require.resolve('babel-loader'),
                                options: Object.assign({
                                    // This is a feature of `babel-loader` for webpack (not Babel itself).
                                    // It enables caching results in ./node_modules/.cache/babel-loader/
                                    // directory for faster rebuilds.
                                    cacheDirectory: true
                                }, babel_server_1.default)
                            },
                            {
                                loader: require.resolve('ts-loader'),
                                options: {
                                    onlyCompileBundledFiles: true,
                                    transpileOnly: true,
                                    happyPackMode: true,
                                    configFile: app_configs_1.default.tsconfig
                                }
                            }
                        ]
                    },
                    // replace css imports with empty files
                    {
                        test: cssRegex,
                        exclude: cssModuleRegex,
                        loader: require.resolve('null-loader')
                    },
                    {
                        test: cssModuleRegex,
                        use: [
                            {
                                loader: require.resolve('css-loader'),
                                options: {
                                    modules: {
                                        exportOnlyLocals: true,
                                        getLocalIdent: getCSSModuleLocalIdent_1.default
                                    },
                                },
                            },
                            {
                                loader: require.resolve('postcss-loader'),
                                options: {
                                    // Necessary for external CSS imports to work
                                    // https://github.com/facebookincubator/create-react-app/issues/2677
                                    ident: 'postcss',
                                    plugins: () => postcss_1.default,
                                },
                            },
                        ],
                    },
                    // "file" loader makes sure those assets get served by WebpackDevServer.
                    // When you `import` an asset, you get its (virtual) filename.
                    // In production, they would get copied to the `build` folder.
                    // This loader doesn't use a "test" so it will catch all modules
                    // that fall through the other loaders.
                    {
                        // Exclude `js` files to keep "css" loader working as it injects
                        // its runtime that would otherwise processed through "file" loader.
                        // Also exclude `html` and `json` extensions so they get processed
                        // by webpacks internal loaders.
                        exclude: [/\.(js|jsx|mjs|ts|tsx)$/, /\.(html|ejs)$/, /\.json$/],
                        loader: require.resolve('file-loader'),
                        options: {
                            outputPath: app_configs_1.default.publicPath,
                            publicPath: app_configs_1.default.publicPath,
                            name: 'static/media/[name].[hash:8].[ext]',
                        },
                    },
                ].filter(Boolean)),
            },
        ],
    },
    plugins: ([
        new webpack_1.default.BannerPlugin({
            banner: assetsIgnoreBanner,
            raw: true,
            entryOnly: false
        }),
        new webpack_1.default.BannerPlugin({
            banner: sourceMapSupportBanner,
            raw: true,
            entryOnly: false
        }),
        app_configs_1.default.tsconfig !== null && new fork_ts_checker_webpack_plugin_1.default()
    ].filter(Boolean)),
});
exports.default = config;
